### PR TITLE
some smaller DefiningOpaqueTypes::No -> Yes switches

### PR DESCRIPTION
**Pull request by oli-obk**
r? @compiler-errors 

These are some easy cases, so let's get them out of the way first.
I added tests exercising the specialization code paths that I believe weren't tested so far.

follow-up to https://github.com/rust-lang/rust/pull/117348

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 1]
```diff
@@ -2595,7 +2601,7 @@ impl<'tcx> SelectionContext<'_, 'tcx> {
                     nested.extend(
                         self.infcx
                             .at(&obligation.cause, obligation.param_env)
-                            .sup(DefineOpaqueTypes::No, source_projection, target_projection)
+                            .sup(DefineOpaqueTypes::Yes, source_projection, target_projection)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:2604]: I don't believe it's correct to include this until we get rid of bubble.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 2]
```diff
@@ -998,7 +1004,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 ty::PredicateKind::Ambiguous => Ok(EvaluatedToAmbig),
                 ty::PredicateKind::Clause(ty::ClauseKind::ConstArgHasType(ct, ty)) => {
                     match self.infcx.at(&obligation.cause, obligation.param_env).eq(
-                        DefineOpaqueTypes::No,
+                        DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1007]: Needs the same comment as above.

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 3]
```diff
@@ -642,7 +642,8 @@ pub fn compute_inherent_assoc_ty_args<'a, 'b, 'tcx>(
         );
     }
 
-    match selcx.infcx.at(&cause, param_env).eq(DefineOpaqueTypes::No, impl_ty, self_ty) {
+    // Inherent associated types are unstable, so we change their behaviour to match the new solver's
+    match selcx.infcx.at(&cause, param_env).eq(DefineOpaqueTypes::Yes, impl_ty, self_ty) {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/project.rs:646]: Yeah, I don't buy this justification, haha.

I don't think we should just change this just "[IATs] are unstable". We shouldn't change this until we change all the other `DefineOpaqueTypes::No` in the solver, which is dependent on us fixing bubble properly.

Can you revert this? Also, even if we wanted to flip this, there's no test associated with this.

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 4]
```diff
@@ -425,7 +425,8 @@ fn plug_infer_with_placeholders<'tcx>(
             if ty.is_ty_var() {
                 let Ok(InferOk { value: (), obligations }) =
                     self.infcx.at(&ObligationCause::dummy(), ty::ParamEnv::empty()).eq(
-                        DefineOpaqueTypes::No,
+                        // Already using new solver, so `DefineOpaqueTypes` is unused
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:428]: Justification for these three are not exactly correct -- we are always unifying with an infer var, so we never expect to see an opaque anyways.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_trait_selection/src/traits/select/mod.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:1007]: ah right, I thought it was only for generic const params, but it's obviously also needed for anon consts in where bounds

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me after rebase

@bors rollup=never (bisect)

**Pull request merged**